plugins/modules/ec2_group.py

```diff
@@ -741,11 +812,27 @@ def ports_expand(ports):
 
 def rule_expand_ports(rule):
     # takes a rule dict and returns a list of expanded rule dicts
+    # uses icmp_code and icmp_type instead of from_ports and to_ports when
+    # available.
     if 'ports' not in rule:
-        if isinstance(rule.get('from_port'), string_types):
-            rule['from_port'] = int(rule.get('from_port'))
-        if isinstance(rule.get('to_port'), string_types):
-            rule['to_port'] = int(rule.get('to_port'))
+        non_icmp_params = any([
+            rule.get('icmp_type', None) is None, rule.get('icmp_code', None) is None])
+        conflict = not non_icmp_params and any([
+            rule.get('from_port', None), rule.get('to_port', None)])
+
+        if non_icmp_params:
+            if isinstance(rule.get('from_port'), string_types):
+                rule['from_port'] = int(rule.get('from_port'))
+            if isinstance(rule.get('to_port'), string_types):
+                rule['to_port'] = int(rule.get('to_port'))
+        else:
+            rule['from_port'] = int(rule.get('icmp_type')) if isinstance(rule.get('icmp_type'), string_types) else rule.get('icmp_type')
+            rule['to_port'] = int(rule.get('icmp_code')) if isinstance(rule.get('icmp_code'), string_types) else rule.get('icmp_code')
+            # Used temporarily to track the fact that icmp keys were converted
+            # to from_port/to_port
+            if not conflict:
+                rule['icmp_keys'] = True
```

Comment:
@Razique Was this "Temporary" in the sense that you were debugging something, or "Temporary" in that it's dropped somewhere I'm missing?

Otherwise LGTM

Reply:
Hey Mark,
so the original author added a set of checks to ensure that users don't pass conflicting parameter in the `validate_rule` function. I have extended the logic to [check against ICMP parameters](https://github.com/ansible-collections/amazon.aws/pull/783/files#diff-2662e61b8cd36633771fa77f2711f866ad852d91d891940c0cfc3c9475f8548dR682).

However:
* That function has no way to know whether the `from_port` and `to_port` keys were added by the module based on `icmp_type` and `icmp_code` as added by the user, or whether they were added by the user, thus having conflicting statements. That `icmp_keys` key allows me to know that this is indeed an internal conversion, thus not complaining if both `icmp_type` or `icmp_code` and `from_port` or `to_port` are present.
* I cannot at this stage simply pop the user keys because subsequent functions read user input before invoking the function. Doing so would have required some more refactoring, and I wanted to limit the scope of change.

If you think it'd be cleaner not to depend on that temporary key, I could instead rework that part.

hope that makes sense :D

Reply:
Thanks, just wanted to make sure that it wasn't dangling code from some testing.